src/async_fuse/memfs/inode.rs

```diff
@@ -0,0 +1,116 @@
+
+#[derive(Debug)]
+pub(crate) struct InodeState {
+    range_begin_end:Mutex<(INum,INum)>,
+    recycle_unused:crossbeam_queue::SegQueue<INum>,
+}
+
+impl InodeState {
+    pub fn new()->Self{
+        Self{
+            range_begin_end:Mutex::new((0,0)),
+            recycle_unused: Default::default(),
+        }
+    }
+}
+
+impl InodeState {
+    /// just get a unique inum
+    async fn alloc_inum(&self, etcd_client:&Arc<EtcdDelegate>) -> INum {
+        const INODE_RANGE:u64=10000;
+        if let Some(inum)= self.recycle_unused.pop(){
+            return inum;
+        }
+        let mut locked=self.range_begin_end.lock().await;
+        if locked.0==locked.1{
+            // need update
+            let ret=etcd::fetch_add_inode_next_range(
+                // here appeared an await, so we need to use tokio::sync::Mutex
+                Arc::clone(etcd_client),INODE_RANGE).await.unwrap_or_else(|e|{
+                panic!("failed to fetch add inode next range, error is {:?}", e)
+            });
+            // begin
+            locked.0=ret;
+            // end
+            locked.1=ret.overflow_add(INODE_RANGE-1);
+            ret
+        }else {
+            let ret=locked.0;
+            locked.0=locked.0.overflow_add(1);
+            ret
+        }
+    }
+
+
+
+    // Check the design at here ：https://github.com/datenlord/datenlord/issues/349
+    //
+    // One case is create new file
+    // Another case is load uncached file from s3
+    // The main point is to avoid two requests with same file path, but different inode number
+
+    // Solution one: yse a global lock for one path, like "lock"+path
+    //  cases:
+    //  1. communicate with other nodes, and got inum
+    //  2. communicate and found no inum;  lock;  communicate and found no inum;  alloc ino;  unlock
+    //  3. communicate and found no inum;  lock;  communicate and found inum;  unlock;
+    //
+    // (chosen) Solution two: use a global inode number for path, like "inode"+path: inum
+    //  cases:
+    //  1. communicate with other nodes, and got inum
+    //  2. communicate and found no inum;  try write kv when there's none;  write success;
+    //  3. communicate and found no inum;  try write kv when there's none;  write failed and get old;
+
+
+    /// get a unique inum for a path when cache miss or creating a new file
+    /// return (inum, is_new)
+    async fn inode_get_inum_by_fullpath(&self,fullpath: &str,node_id:&str,volume_info:&str,etcd_client:&Arc<EtcdDelegate>) -> (INum,bool){
+        //  1. communicate with other nodes and try to get existing inode info
+        let fattr=dist::client::get_attr(
+            Arc::clone(etcd_client),node_id,volume_info,fullpath).await.unwrap_or_else(|e|{
+            panic!("failed to get attr, error is {:?}", e)
+        });
+        match fattr {
+            None => {
+                let inum=self.alloc_inum(etcd_client).await;
+                // try write kv when there's none
```

Comment:
Please don't repeat the code logic in the comment, try to explain the reason instead.